state/application.go

```diff
@@ -1005,6 +1005,108 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 	return nil
 }
 
+// UpdateApplicationSeries updates the series for the Application.
+func (a *Application) UpdateApplicationSeries(series string, force bool) (err error) {
+	acopy := &Application{a.st, a.doc}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		a := acopy
+		if attempt > 0 {
+			// If we've tried once already and failed, re-evaluate the criteria.
+			if err := a.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+
+		var ops []txn.Op
+
+		appOps, err := a.addApplicationUpdateSeriesOps(series, force)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops = append(ops, appOps...)
+
+		units, err := a.AllUnits()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if len(units) > 0 {
+			// All units have the same subordinates...
+			u := units[0]
+			subNames := u.SubordinateNames()
+			ops = append(ops, txn.Op{
+				C:      unitsC,
+				Id:     u.doc.DocID,
+				Assert: bson.D{{"subordinates", subNames}},
+			})
+			ops = append(ops, txn.Op{
+				C:      unitsC,
+				Id:     u.doc.DocID,
+				Assert: txn.DocExists,
+			})
+
+			for _, n := range subNames {
+				sub := strings.Split(n, "/")[0]
+				subApp, err := a.st.Application(sub)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				appOps, err := subApp.addApplicationUpdateSeriesOps(series, force)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				ops = append(ops, appOps...)
+			}
+		}
+		return ops, nil
+	}
+
+	err = a.st.run(buildTxn)
+	if err != nil && IsIncompatibleSeriesError(err) {
+		return errors.Annotatef(err, "cannot update series for %q to %s", a, series)
+	}
+	return err
+}
+
+func (a *Application) addApplicationUpdateSeriesOps(series string, force bool) ([]txn.Op, error) {
+	var ops []txn.Op
+	err := a.VerifySupportedSeries(series, force)
```

Comment:
This is done outside the ops funcs. And then the ops func build up asserts to match.